packages/vite/src/node/plugins/define.ts

```diff
@@ -157,7 +158,28 @@ export async function replaceDefine(
     sourcemap: config.command === 'build' ? !!config.build.sourcemap : true,
   })
 
+  // remove esbuild's <define:...> source entries
+  // since they would confuse source map remapping/collapsing which expects a single source
+  if (result.map.includes('<define:')) {
+    const originalMap = new TraceMap(result.map)
+    if (originalMap.sources.length >= 2) {
+      const sourceIndex = originalMap.sources.indexOf(id)
+      const decoded = decodedMap(originalMap)
+      decoded.sources = [id]
+      decoded.mappings = decoded.mappings.map((segments) =>
+        segments.filter((segment) => {
+          // modify and filter
+          const index = segment[1]
+          segment[1] = 0
+          return index === sourceIndex
+        }),
+      )
+      result.map = JSON.stringify(encodedMap(new TraceMap(decoded)))
+    }
+  }
+
   for (const marker in replacementMarkers) {
+    // TODO: this could also break sourcemap?
     result.code = result.code.replaceAll(marker, replacementMarkers[marker])
```

Comment:
I thought this could be also the cause of source map breakage, but I expect this won't affect the validity of source map for most of the cases since this only changes column positions where `import.meta.env` appears:

```ts
console.log('test', import.meta.env)
// 
//  ⇓ esbuild with define: { "import.meta.env": "____very_long_hash____" }
// 
console.log('test', ____very_long_hash____)
// 
//  ⇓ replaceAll(marker, ...)
// 
console.log('test', { "BASE_URL": "/", ... })
```

So, I suppose bringing `magic-string` here to patch up sourcemap would be probably overkill.